File: DatastoreConnection.ts
import DataStore2 from "@rbxts/datastore2";

export interface IDatastoreConnection<T, DSKeyType extends string = string> {
	key: DSKeyType;
	default: T;
}

export default class DatastoreConnection<T, serialized = T, DSKeyType extends string = string>
	implements IDatastoreConnection<T, DSKeyType>
{
	key: DSKeyType;
	default: T;

	player: Player;

	private isTable: boolean;

	// Default: Doesn't change anything
	private serializeFunction?: (obj: T) => serialized;

	// Default: Doesn't change anything
	private deserializeFunction?: (obj: serialized) => T;

	constructor(
		prop: IDatastoreConnection<T, DSKeyType>,
		player: Player,
		isTable: boolean,
		serializeFunction?: (obj: T) => serialized,
		deserializeFunction?: (obj: serialized) => T,
	) {
		this.key = prop.key;
		this.default = prop.default;

		this.player = player;
		this.isTable = isTable;

		if (serializeFunction) this.serializeFunction = serializeFunction;
		if (deserializeFunction) this.deserializeFunction = deserializeFunction;
	}

	public getDatastore(): T {
		if (this.deserializeFunction && this.serializeFunction) {
			return this.deserializeFunction(
				DataStore2<serialized>(this.key, this.player).Get(this.serializeFunction(this.default)),
			);
		} else if (this.isTable) {
			return DataStore2<T>(this.key, this.player).GetTable(<object>(<unknown>this.default));
		} else return DataStore2<T>(this.key, this.player).Get(this.default);
	}

	public setDatastore(newData: T): void {
		if (this.serializeFunction) {
			const serializedData = this.serializeFunction(newData);
			DataStore2<serialized>(this.key, this.player).Set(serializedData);
		} else DataStore2<T>(this.key, this.player).Set(newData);
	}
}

File: DatastoreConnection.lua
-- Compiled with roblox-ts v1.3.3
local TS = require(game:GetService("ReplicatedStorage"):WaitForChild("rbxts_include"):WaitForChild("RuntimeLib"))
local DataStore2 = TS.import(script, TS.getModule(script, "@rbxts", "datastore2").src)
local DatastoreConnection
do
	DatastoreConnection = setmetatable({}, {
		__tostring = function()
			return "DatastoreConnection"
		end,
	})
	DatastoreConnection.__index = DatastoreConnection
	function DatastoreConnection.new(...)
		local self = setmetatable({}, DatastoreConnection)
		return self:constructor(...) or self
	end
	function DatastoreConnection:constructor(prop, player, isTable, serializeFunction, deserializeFunction)
		self.key = prop.key
		self.default = prop.default
		self.player = player
		self.isTable = isTable
		if serializeFunction then
			self.serializeFunction = serializeFunction
		end
		if deserializeFunction then
			self.deserializeFunction = deserializeFunction
		end
	end
	function DatastoreConnection:getDatastore()
		if self.deserializeFunction and self.serializeFunction then
			return self.deserializeFunction(DataStore2(self.key, self.player):Get(self.serializeFunction(self.default)))
		elseif self.isTable then
			return DataStore2(self.key, self.player):GetTable((self.default))
		else
			return DataStore2(self.key, self.player):Get(self.default)
		end
	end
	function DatastoreConnection:setDatastore(newData)
		if self.serializeFunction then
			local serializedData = self.serializeFunction(newData)
			DataStore2(self.key, self.player):Set(serializedData)
		else
			DataStore2(self.key, self.player):Set(newData)
		end
	end
end
return {
	default = DatastoreConnection,
}
